src/server/ua_server_binary.c

```diff
@@ -996,3 +997,202 @@ UA_Server_networkCallback(UA_ConnectionManager *cm, uintptr_t connectionId,
+        return retval;
+    }
+
+    UA_LOG_DEBUG(&server->config.logger, UA_LOGCATEGORY_CLIENT, "Sent RHE message");
+    return UA_STATUSCODE_GOOD;
+}
+
+static void retryReverseConnectCallback(UA_Server *server, void *context) {
+    (void)context;
+
+    reverse_connect_context *rc = NULL;
+    SLIST_FOREACH(rc, &server->reverseConnects, next) {
+        if (rc && rc->status == REVERSE_CONNECT_STATUS_DISCONNECTED) {
+            UA_LOG_INFO(&server->config.logger, UA_LOGCATEGORY_SERVER,
+                        "Attempt to reverse reconnect to %.*s:%d", (int)rc->hostname.length,
+                        rc->hostname.data, rc->port);
+            UA_Server_attemptReverseConnect(server, rc);
+        }
+    }
+}
+
+UA_StatusCode UA_Server_setReverseConnectRetryCallback(UA_Server *server, UA_Boolean enabled) {
+    if (enabled && !server->reverseConnectsCheckHandle) {
+        const UA_UInt32 reconnectInterval = server->config.reverseReconnectInterval ?
+                    server->config.reverseReconnectInterval : 15000;
+        return UA_Server_addRepeatedCallback(server, retryReverseConnectCallback, NULL,
+                                             reconnectInterval, &server->reverseConnectsCheckHandle);
+    } else if (!enabled && server->reverseConnectsCheckHandle) {
+        UA_Server_removeCallback(server, server->reverseConnectsCheckHandle);
+        server->reverseConnectsCheckHandle = 0;
+        return UA_STATUSCODE_GOOD;
+    }
+
+    return UA_STATUSCODE_GOOD;
+}
+
+void UA_Server_reverseConnectCallback(UA_ConnectionManager *cm, uintptr_t connectionId,
+                          void *application, void **connectionContext,
+                          UA_ConnectionState state,
+                          size_t paramsSize, const UA_KeyValuePair *params,
+                          UA_ByteString msg) {
+    (void)paramsSize;
+    (void)params;
+
+    UA_Server *server = (UA_Server*)application;
+
+    /* UA_LOG_INFO(&server->config.logger, UA_LOGCATEGORY_SERVER, "Reverse connect has state %d", state); */
+
+    reverse_connect_context *context = (reverse_connect_context *)*connectionContext;
+
+    if(context->currentConnection == NULL) {
+        /* The socket is closing without being previously registered -> ignore */
+        if(state == UA_CONNECTIONSTATE_CLOSED ||
+           state == UA_CONNECTIONSTATE_CLOSING) {
+            context->status = REVERSE_CONNECT_STATUS_DISCONNECTED;
+            return;
+        }
+
+        /* Cannot register */
+        if(server->serverConnectionsSize >= UA_MAXSERVERCONNECTIONS) {
+            UA_LOG_WARNING(&server->config.logger, UA_LOGCATEGORY_SERVER,
+                           "Cannot register server socket - too many already open");
+            cm->closeConnection(cm, connectionId);
+            context->status = REVERSE_CONNECT_STATUS_DISCONNECTED;
+            return;
+        }
+
+        /* Find and use a free connection slot */
+        server->serverConnectionsSize++;
```

Comment:
The ServerConnections are for listening sockets.
Not connections to individual clients.

Reply:
We already distinguish between listen-sockets and "active connections".
Let's not add a third kind. To me, reverse connections are also "active connections".
You can immediately create a SecureChannel to store the state. And add new status in the SecureChannelState enum.

Reply:
@jpfr If I understand you correctly, I should add a secure channel to server->channels and extend the UA_SecureChannel struct with the necessary information (isReverseConnect flag, URL for reverse connect, etc...)?